src/index.ts

```diff
@@ -1,12 +1,13 @@
 // https://github.com/libp2p/js-libp2p/issues/1269#issue-1280763716
-export class CodeError<T extends Record<string, any> = Record<string, never>> extends Error {
-  code: string
-  props: T
-
-  constructor (message: string, code: string, props?: T) {
+export class CodeError<T extends Readonly<Record<string, any>>> extends Error {
+  constructor (
+    readonly message: string,
+    readonly code: string,
+    readonly props: T = {} as T // eslint-disable-line @typescript-eslint/consistent-type-assertions
```

Comment:
I think this won't give the right type in the default case. Probably should keep the default generic type and set props in the constructor body.

```ts
const err = new CodeError('foo', 'bar');
err.props // <- what is this type? I think ideally it should be Record<string, never>
```

As far as `readonly props` is concerned, I think that's fine. Any later additions to the props would not be able to replace props entirely, but could rather replace individual fields.

```ts
interface SomeProps {
  a: string
}
const err = new CodeError('foo', 'bar', {a: ''} as SomeProps)
...
err.props.a = 'a real value' // <- still allowed even tho props is readonly
```

Reply:
> err.props // <- what is this type? I think ideally it should be Record<string, never>

it ends up being Readonly<Record<string, any>> which gives very similar characteristics. I guess when extends is being used, and the generic type and parameter are not supplied, the type becomes whatever the generic was extending.

Reply:
> Probably should keep the default generic type and set props in the constructor body.

I like how this looks right now but both are the same so will squash with just agreed changes if you have the same thoughts when I wake up tomorrow.

agreed:
- message, code, and props are readonly
- message and code assigned inside parameters

disagreed:
- assigning props inside parameters
- default generic type vs Readonly utility

> Any later additions to the props would not be able to replace props entirely, but could rather replace individual fields.

:+1: